packages/blob/src/api.ts

```diff
@@ -0,0 +1,171 @@
+}
+
+// This version is used to ensure that the client and server are compatible
+// The server (Vercel Blob API) uses this information to change its behavior like the
+// response format
+const BLOB_API_VERSION = 6;
+
+function getApiVersion(): string {
+  let versionOverride = null;
+  try {
+    // wrapping this code in a try/catch as this function is used in the browser and Vite doesn't define the process.env.
+    // As this varaible is NOT used in production, it will always default to the BLOB_API_VERSION
+    versionOverride =
+      process.env.VERCEL_BLOB_API_VERSION_OVERRIDE ||
+      process.env.NEXT_PUBLIC_VERCEL_BLOB_API_VERSION_OVERRIDE;
+  } catch {
+    // noop
+  }
+
+  return `${versionOverride ?? BLOB_API_VERSION}`;
+}
+
+function getApiUrl(pathname = ''): string {
+  let baseUrl = null;
+  try {
+    // wrapping this code in a try/catch as this function is used in the browser and Vite doesn't define the process.env.
+    // As this varaible is NOT used in production, it will always default to production endpoint
+    baseUrl =
+      process.env.VERCEL_BLOB_API_URL ||
+      process.env.NEXT_PUBLIC_VERCEL_BLOB_API_URL;
+  } catch {
+    // noop
+  }
+  return `${baseUrl || 'https://blob.vercel-storage.com'}${pathname}`;
+}
+
+function getRetries(): number {
+  try {
+    const retries = process.env.VERCEL_BLOB_RETRIES || '10';
+
+    return parseInt(retries, 10);
+  } catch {
+    return 10;
+  }
+}
+
+async function validateBlobApiResponse(response: Response): Promise<void> {
+  if (!response.ok) {
+    let data: unknown;
+
+    try {
+      data = await response.json();
+    } catch {
+      throw new BlobUnknownError();
+    }
+
+    const error = (data as BlobApiError).error;
+
+    switch (error?.code) {
+      case 'store_suspended':
+        throw new BlobStoreSuspendedError();
+      case 'forbidden':
+        throw new BlobAccessError();
+      case 'not_found':
+        throw new BlobNotFoundError();
+      case 'store_not_found':
+        throw new BlobStoreNotFoundError();
+      case 'bad_request':
+        throw new BlobError(error.message ?? 'Bad request');
+      case 'service_unavailable':
+        throw new BlobServiceNotAvailable();
+      case 'unknown_error':
+      case 'not_allowed':
+      default:
+        throw new BlobUnknownError();
+    }
+  }
+}
+
+export async function requestApi<TResponse>(
+  pathname: string,
+  init: RequestInit,
+  commandOptions: BlobCommandOptions | undefined,
+): Promise<TResponse> {
+  const apiResponse = await retry(
+    async () => {
+      const res = await fetch(getApiUrl(pathname), {
+        ...init,
+        headers: {
+          'x-api-version': getApiVersion(),
+          authorization: `Bearer ${getTokenFromOptionsOrEnv(commandOptions)}`,
+
+          ...init.headers,
+        },
+      });
+
+      if (res.status >= 500) {
+        // this will be retried and shown to the user if no more retries are left
+        await validateBlobApiResponse(res);
+      }
```

Comment:
shouldn't we bail on known errors? all except `unknown_error` and `unknown_error`, no?

Reply:
true

Reply:
the [async-retry](https://www.npmjs.com/package/async-retry) package offers a `bail` function to be able to throw an error without retrying it. I think we need something like
```
if (code === 'unknown_error' || code === 'service_unavailable') {
   // retry
   throw getBlobError(error);
} if (code) {
  // do NOT retry
  bail(getBlobError(error))
}
```
otherwise we'd return a response containing the error

Reply:
it would be nice to have a unit test for this code, as I'm not completely sure how it behaves

Reply:
ah thats nice. It work currently because i check. the response outside of the retry again. But I can remove that than

Reply:
added some tests